sql/set_var.cc

```diff
@@ -733,6 +735,61 @@ sys_var *intern_find_sys_var(const char *str, size_t length) {
   return var;
 }
 
+/**
+  Helper function to set variables.
+  Use the thd -> list<var> map to iterate over all threads
+  For each thread, for each variable in its list, perform the specified function
+  @param THD            Current thread
+  @param thd_var_map    Map of threads to list of variables to be updated
+  @param func           Function to be executed for each variable
+  @retval
+    0   ok
+  @retval
+    1   ERROR
+*/
+
+static int process_set_variable_map(THD *thd, set_var_list_map &thd_var_map,
+                                    int (*func)(THD *, set_var_base *)) {
+  DBUG_ENTER("process_set_variable_map");
+  int error = 0;
+
+  for (auto var_iter : thd_var_map) {
+    ulong thd_id = var_iter.first;
+    THD *change_thd = thd;
+    // Thread id of the current thread is 0
+    if (thd_id) {
+      Find_thd_with_id find_thd_with_id(thd_id);
+      change_thd =
+          Global_THD_manager::get_instance()->find_thd(&find_thd_with_id);
+      if (!change_thd) {
+        my_error(ER_NO_SUCH_THREAD, MYF(0), thd_id);
+        DBUG_RETURN(1);
+      }
+
+      Security_context *sctx = thd->security_context();
+      if (!sctx->check_access(SUPER_ACL) &&
+          !sctx->has_global_grant(STRING_WITH_LEN("SYSTEM_VARIABLES_ADMIN"))
+               .first) {
+        mysql_mutex_unlock(&change_thd->LOCK_thd_data);
+        my_error(ER_SHOW_DENIED_ERROR, MYF(0), thd_id);
```

Comment:
I don't know if the change of behaviour here is intentional, but previously, we just checked if the user was an owner of the thread, instead of checking for specific privileges. Now the error is kind of misleading, because you'll get "You are not owner of thread", when we should really be getting something from `ER_SPECIFIC_ACCESS_DENIED_ERROR`.

Reply:
I spoke with @hermanlee and the change in behaviour is probably okay.

However, I think we should use `ER_SPECIFIC_ACCESS_DENIED_ERROR` instead so that the error is less confusing.

Reply:
Done